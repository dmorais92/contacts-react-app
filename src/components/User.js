import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import logo from "../logo.svg";
import Card from "./common/Card";

import "./User.scss";
import { chunk, capitalize } from "../utils";

const renderGeoFields = coords => {
  const { lat, lng } = coords;
  return (
    (lat && lng && (
      <div className="geo-fields title-value" key="geo">
        <a
          target="blank"
          href={`https://www.google.com/maps/search/?api=1&query=${lat},${lng}`}
        >
          <h5>Location</h5>
          <span>
            {lat}ยบ, {lng}ยบ
          </span>
        </a>
      </div>
    )) ||
    ""
  );
};

class UserList extends PureComponent {
  static propTypes = {
    user: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
      username: PropTypes.string,
      address: PropTypes.object,
      phone: PropTypes.string,
      website: PropTypes.string,
      company: PropTypes.object
    })
  };

  renderUserFields(fields, target = this.props.user) {
    if (fields && fields.length) {
      return chunk(fields, 3).map(section => (
        <div className="info-section" key={`${fields}-${section}`}>
          {section.map(field => {
            switch (field) {
              case "geo":
                return renderGeoFields(target[field]);
              default:
                return (
                  <div
                    className="title-value"
                    key={`${field}-${target[field]}`}
                  >
                    <h5>{capitalize(field)}</h5>
                    <span>{target[field]}</span>
                  </div>
                );
            }
          })}
        </div>
      ));
    }
  }

  render() {
    const { user } = this.props;
    if (user) {
      const { name, id, username, email, address, company, ...userInfo } = user;
      const extraFields = userInfo && Object.keys(userInfo);

      return (
        <Card vertical key={id} classes={["User"]}>
          <div className="row">
            <img src={logo} className="logo" alt="logo" />
            <div className="main-info">
              <h3>{name}</h3>
              <span>{email}</span>
              <span>{username}</span>
            </div>
          </div>
          {this.renderUserFields(extraFields)}
          {this.renderUserFields(Object.keys(address), address)}
        </Card>
      );
    }
    return '';
  }
}

export default UserList;
