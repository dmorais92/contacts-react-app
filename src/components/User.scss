@import '../variables.scss';

.User {
  .row {
    display: flex;
    flex-flow: row nowrap;
    justify-content: space-between;
    align-items: center;
    .main-info {
      display: flex;
      flex-flow: column wrap;
      h3 {
        margin: 0;
      }
    }
  }
  .info-section {
    flex-flow: column nowrap;
    align-items: flex-end;
    border-bottom: 1px solid $secondary-font-color;
    margin-left: 10vh;
    min-width: 60%;
    .title-value {
      flex-flow: row nowrap;
      align-items: space-between;
      a {
        text-decoration: none;
        background-color: #eee;
        &:hover {
          text-decoration: underline;
        }
      }
      h5 {
        margin: 0.25em 0;
      }
      span {
        margin: 0 0.25em;
      }
    }
  }
}
