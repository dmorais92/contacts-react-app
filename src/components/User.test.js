import React from 'react';
import ReactDOM from 'react-dom';
import User from './User';
import renderer from 'react-test-renderer';

const TEST_USER =   {
  "id": 1,
  "name": "Leanne Graham",
  "username": "Bret",
  "email": "Sincere@april.biz",
  "address": {
    "street": "Kulas Light",
    "suite": "Apt. 556",
    "city": "Gwenborough",
    "zipcode": "92998-3874",
    "geo": {
      "lat": "-37.3159",
      "lng": "81.1496"
    }
  },
  "phone": "1-770-736-8031 x56442",
  "website": "hildegard.org",
  "company": {
    "name": "Romaguera-Crona",
    "catchPhrase": "Multi-layered client-server neural-net",
    "bs": "harness real-time e-markets"
  }
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<User />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders a user', () => {
  const component = renderer.create(
    <User user={TEST_USER}/>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
