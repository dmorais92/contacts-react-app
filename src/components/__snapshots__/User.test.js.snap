// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a user 1`] = `
<div
  className="Card vertical User"
>
  <div
    className="row"
  >
    <img
      alt="logo"
      className="logo"
      src="logo.svg"
    />
    <div
      className="main-info"
    >
      <h3>
        Leanne Graham
      </h3>
      <span>
        Sincere@april.biz
      </span>
      <span>
        Bret
      </span>
    </div>
  </div>
  <div
    className="info-section"
  >
    <div
      className="title-value"
    >
      <h5>
        Phone
      </h5>
      <span>
        1-770-736-8031 x56442
      </span>
    </div>
    <div
      className="title-value"
    >
      <h5>
        Website
      </h5>
      <span>
        hildegard.org
      </span>
    </div>
  </div>
  <div
    className="info-section"
  >
    <div
      className="title-value"
    >
      <h5>
        Street
      </h5>
      <span>
        Kulas Light
      </span>
    </div>
    <div
      className="title-value"
    >
      <h5>
        Suite
      </h5>
      <span>
        Apt. 556
      </span>
    </div>
    <div
      className="title-value"
    >
      <h5>
        City
      </h5>
      <span>
        Gwenborough
      </span>
    </div>
  </div>
  <div
    className="info-section"
  >
    <div
      className="title-value"
    >
      <h5>
        Zipcode
      </h5>
      <span>
        92998-3874
      </span>
    </div>
    <div
      className="geo-fields title-value"
    >
      <a
        href="https://www.google.com/maps/search/?api=1&query=-37.3159,81.1496"
        target="blank"
      >
        <h5>
          Location
        </h5>
        <span>
          -37.3159
          ยบ, 
          81.1496
          ยบ
        </span>
      </a>
    </div>
  </div>
</div>
`;
